store/src/java/com/zimbra/cs/service/util/SecretKey.java

```diff
@@ -0,0 +1,81 @@
+package com.zimbra.cs.service.util;
+
+import com.google.common.base.Strings;
+import com.zimbra.common.service.ServiceException;
+import com.zimbra.cs.account.Provisioning;
+import org.apache.commons.codec.binary.Hex;
+
+import javax.mail.MessagingException;
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.Base64;
+
+public class SecretKey {
+
+    public static final int KEY_SIZE_BYTES = 32;
+    public static final String MSGVRFY_HEADER_PREFIX = "hash=SHA256;guid=";
+    public static final String MSGVRFY_ALGORITHM_NAME = "SHA-256";
+
+    /**
+     * returns the randomly generated String
+     * @return randomly generated String
+     * @throws ServiceException if an error occurred
+     */
+    public static String generateRandomString() throws ServiceException {
+        try {
+            SecureRandom random = SecureRandom.getInstance("SHA1PRNG");
+            byte[] key = new byte[KEY_SIZE_BYTES];
+            random.nextBytes(key);
+            return new String(Hex.encodeHex(key));
+        } catch (NoSuchAlgorithmException e) {
+            throw ServiceException.FAILURE("unable to initialize SecureRandom for mail recall", e);
+        }
+    }
+
+    /**
+     * provide the Hash for Message-Verification header field.
+     * @param id
+     * @param date
+     * @param from
+     * @throws ServiceException
+     */
+    public static String getMessageVerificationHeaderValue(String id, String date, String from) throws MessagingException, ServiceException {
+        String secretKey = Provisioning.getInstance().getConfig().getSecretKeyForMailRecall();
+        if(Strings.isNullOrEmpty(secretKey)) {
+            secretKey = SecretKey.generateRandomString();
+            Provisioning.getInstance().getConfig().setSecretKeyForMailRecall(secretKey);
```

Comment:
Multiple server environment need to be considered.
What occurs when the attribute is empty and 2 or more mailstore servers execute sending message process at the same time? A different value is generated on each mailstore, and then the different value is stored in each cache store until the cache is expired or mailboxd is restarted.
To make it more stable,
1. generate a key on one of mailstore servers
2. run flushcache config on all servers
3. enable recall function with the single key - without any conflict/mismatch among mailstore servers.

Reply:
Done